tsl/src/remote/copy_fetcher.c

```diff
@@ -556,35 +532,49 @@ copy_fetcher_close(DataFetcher *df)
 	 */
 	if (fetcher->state.open && fetcher->state.data_req != NULL)
 	{
-		int end_res = PQendcopy(remote_connection_get_pg_conn(fetcher->state.conn));
-		if (end_res != 0)
+		PGconn *conn = remote_connection_get_pg_conn(fetcher->state.conn);
+		PGresult *final_pgres = NULL;
+		PGresult *pgres = NULL;
+		unsigned int num_results = 0;
+
+		pfree(fetcher->state.data_req);
+		fetcher->state.data_req = NULL;
+
+		if (!fetcher->state.eof)
 		{
-			TSConnectionError err;
-			remote_connection_get_error(fetcher->state.conn, &err);
-			remote_connection_error_elog(&err, ERROR);
+			copy_fetcher_reset(fetcher);
+			elog(ERROR, "COPY fetcher closed without reaching EOF");
 		}
 
-		/*
-		 * Shouldn't have any activity on the connection after we have
-		 * finished COPY. Just double-check.
-		 */
-		PGresult *pgres = PQgetResult(remote_connection_get_pg_conn(fetcher->state.conn));
-		if (pgres)
+		/* Read results until NULL */
+		while ((pgres = PQgetResult(conn)))
+		{
+			if (final_pgres == NULL)
+				final_pgres = pgres;
+			else
+				remote_result_close(pgres);
+			num_results++;
+		}
+
+		Assert(num_results == 1);
```

Comment:
I'd make this an `elog` because this is closer to an environment error, we don't control the remote server.

Reply:
Probably would also be good to see what other PQResults we got instead of the single one we expected. Although I'm not sure if this ever happens.

Reply:
There's a check above for eof being set, so we shouldn't get here if the connection isn't in the right state. This is what the assert checks. For tests, it is good to crash here for something that shouldn't happen.